CONTRIBUTING.md

```diff
@@ -6,6 +6,212 @@ This guide explains how to contribute to cocotb, and documents the processes we
+
+* Cocotb's core functionality
+* How to write Verilog and VHDL
+* How to write cocotb tests in Python
+* (Optionally) [codecov](https://docs.codecov.io/docs); coverage aggregator and Github bot
+* (Optionally) the [coverage](https://coverage.readthedocs.io/en/latest/) module, for Python code coverage
+* (Optionally) [gcov](https://gcc.gnu.org/onlinedocs/gcc/Gcov.html), for C++ code coverage
+* (Optionally) [Github Actions](https://docs.github.com/en/free-pro-team@latest/actions), for automatic acceptance testing
+
+Cocotb's regression tests can be improved by:
+
+* Testing more of cocotb's core functionality
+* Testing corner cases left out of the current set of tests (identified by looking at code coverage)
+* Increasing the matrix of simulators, operating system, and Python installations tested in CI
+
+Testing improvements don't require maintainer pre-approval, but require a pull request.
+Please see the [guidelines on submitting pull requests](#patch-requirements).
+
+
+### Features
+
+Cocotb is still in development and new features are still welcome and appreciated;
+as long as they stay [in scope](#Architecture-and-Scope-of-Cocotb).
+Cocotb is comprised of several major codebases, each requiring different sets of skills and development process.
+Instead of including that breakdown here, it is done in the [internal documentation](https://github.com/cocotb/cocotb/wiki/cocotb-Internals).
+
+Small improvements to existing features generally do not require maintainer pre-approval.
+Large changes, approximately >150 LOC changed, and new features generally require maintainer pre-approval.
+
+***New features must not break existing features.***
+
+Feature changes require full coverage of the added feature.
+This likely requires adding new unit tests to the `tests` directory.
+Issue-specific test directories will not be accepted, unless a special HDL entity is required.
+Instead, place the test in an existing test suite (test_cocotb, test_discovery, etc.).
+
+Features should generally follow the following design principles:
+
+* Something the user cannot do without assistance of cocotb-specific code
+* Orthogonal to existing features
+* Easily composed with existing features
+* Limited in scope and impervious to scope creep
+
+
+### Bugfixes
+
+**!WARNING!** Bugfixing cocotb is not for the faint of heart!
+
+Bugs happen.
+cocotb supports many simulators that have inconsistent support for the procedural interfaces cocotb depends on,
+and it has a number of features that aren't wholly tested yet.
+There are likely many bugs lurking, waiting to be found;
+which is why increasing testing and code coverage is important.
+Working on bugfixing can be very challenging, depending on the cause of the bug.
+In general, bugfixing requires:
+
+* How cocotb works
+* cocotb's debugging utilities
+* (Optional) Simulator interfaces (VPI, VHPI, and FLI)
+* (Optional) Python debugging tools (pdb, cherrypy)
+* (Optional) C/C++ debugging tools (gdb, valgrind)
```

Comment:
Links into Wiki on how to use each of these tools. Will create stubs at a minimum before finishing the PR.